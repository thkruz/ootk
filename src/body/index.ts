export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
