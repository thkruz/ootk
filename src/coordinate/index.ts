export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
