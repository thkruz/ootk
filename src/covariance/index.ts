export * from './CovarianceSample.js';
export * from './StateCovariance.js';
