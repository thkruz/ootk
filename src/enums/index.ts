export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
