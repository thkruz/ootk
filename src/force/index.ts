export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

