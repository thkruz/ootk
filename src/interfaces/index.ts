export type { BaseObjectParams } from './BaseObjectParams.js';
export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { SatelliteParams } from './SatelliteParams.js';
export type { SensorParams } from './SensorParams.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { RfSensorParams } from './RfSensorParams.js';
