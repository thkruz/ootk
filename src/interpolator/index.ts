export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

