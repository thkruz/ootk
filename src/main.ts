/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
