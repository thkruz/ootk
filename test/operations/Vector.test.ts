import { DEG2RAD, Radians, Vector } from './../../src/main';

describe('Vector', () => {
  // create a Vector with elements and get its length
  it('should create a Vector with elements and get its length', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.length).toBe(3);
  });

  // create a Vector with all elements set to zero and get its length
  it('should create a Vector with all elements set to zero and get its length', () => {
    const v = Vector.zero(3);

    expect(v.length).toBe(3);
  });

  // create a Vector with all elements set to a value and get its length
  it('should create a Vector with all elements set to a value and get its length', () => {
    const v = Vector.filled(3, 5);

    expect(v.length).toBe(3);
  });

  // create a Vector from a list of elements and get its length
  it('should create a Vector from a list of elements and get its length', () => {
    const v = Vector.fromList([1, 2, 3]);

    expect(v.length).toBe(3);
  });

  // get the x, y, and z components of a Vector
  it('should get the x, y, and z components of a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.x).toBe(1);
    expect(v.y).toBe(2);
    expect(v.z).toBe(3);
  });

  // get the magnitude of a Vector
  it('should get the magnitude of a Vector', () => {
    const v = new Vector([3, 4]);

    expect(v.magnitude()).toBe(5);
  });

  // create a Vector with an empty array of elements
  it('should create a Vector with an empty array of elements', () => {
    const v = new Vector([]);

    expect(v.length).toBe(0);
  });

  // create a Vector with a negative length
  it('should create a Vector with a negative length', () => {
    expect(() => new Vector([-1, -2, -3])).toMatchSnapshot();
  });

  // add two Vectors and return a new Vector
  it('should add two Vectors and return a new Vector', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);
    const result = v1.add(v2);

    expect(result).toMatchSnapshot();
  });

  // subtract two Vectors and return a new Vector
  it('should subtract two Vectors and return a new Vector when called', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);
    const result = v1.subtract(v2);

    expect(result).toMatchInlineSnapshot(`
      Vector {
        "elements": Array [
          -3,
          -3,
          -3,
        ],
        "length": 3,
      }
    `);
  });

  // scale a Vector by a scalar and return a new Vector
  it('should scale a Vector by a scalar and return a new Vector when called with a scalar value', () => {
    const v = new Vector([1, 2, 3]);
    const scaledVector = v.scale(2);

    expect(scaledVector).toMatchInlineSnapshot(`
      Vector {
        "elements": Array [
          2,
          4,
          6,
        ],
        "length": 3,
      }
    `);
  });

  // negate a Vector and return a new Vector
  it('should negate a Vector and return a new Vector when calling the negate() method', () => {
    const v = new Vector([1, 2, 3]);
    const negated = v.negate();

    expect(negated).toEqual(new Vector([-1, -2, -3]));
  });

  // normalize a Vector and return a new Vector
  it('should normalize a Vector and return a new Vector', () => {
    const v = new Vector([3, 4]);
    const normalized = v.normalize();

    expect(normalized).toMatchSnapshot();
  });

  // calculate the dot product of two Vectors
  it('should calculate the dot product of two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);
    const dotProduct = v1.dot(v2);

    expect(dotProduct).toMatchSnapshot();
  });

  // toString
  it('should return a string representation of a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.toString()).toMatchSnapshot();
  });

  // toList
  it('should return a list representation of a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.toList()).toMatchSnapshot();
  });

  // toArray
  it('should return an array representation of a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.toArray()).toMatchSnapshot();
  });

  // distance
  it('should calculate the distance between two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.distance(v2)).toMatchSnapshot();
  });

  // outer
  it('should calculate the outer product of two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.outer(v2)).toMatchSnapshot();
  });

  // cross
  it('should calculate the cross product of two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.cross(v2)).toMatchSnapshot();
  });

  // skewSymmetric
  it('should calculate the skew symmetric matrix of a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.skewSymmetric()).toMatchSnapshot();
  });

  // rotX
  it('should rotate a Vector around the x-axis', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.rotX((90 * DEG2RAD) as Radians)).toMatchSnapshot();
  });

  // rotY
  it('should rotate a Vector around the y-axis', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.rotY((90 * DEG2RAD) as Radians)).toMatchSnapshot();
  });

  // rotZ
  it('should rotate a Vector around the z-axis', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.rotZ((90 * DEG2RAD) as Radians)).toMatchSnapshot();
  });

  // angle
  it('should calculate the angle between two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.angle(v2)).toMatchSnapshot();
  });

  // angleDegrees
  it('should calculate the angle between two Vectors in degrees', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.angleDegrees(v2)).toMatchSnapshot();
  });

  // sight
  it('should calculate the sight of a Vector', () => {
    const v = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v.sight(v2, (90 * DEG2RAD) as Radians)).toMatchSnapshot();
  });

  // bisect
  it('should calculate the bisect of two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.bisect(v2)).toMatchSnapshot();
  });

  // slice
  it('should slice a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.slice(1, 2)).toMatchSnapshot();
  });

  // join
  it('should join two Vectors', () => {
    const v1 = new Vector([1, 2, 3]);
    const v2 = new Vector([4, 5, 6]);

    expect(v1.join(v2)).toMatchSnapshot();
  });

  // row
  it('should get a row from a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.row()).toMatchSnapshot();
  });

  // column
  it('should get a column from a Vector', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.column()).toMatchSnapshot();
  });

  // toVector3D
  it('should convert a Vector to a Vector3D', () => {
    const v = new Vector([1, 2, 3]);

    expect(v.toVector3D(1)).toMatchSnapshot();
  });
});
