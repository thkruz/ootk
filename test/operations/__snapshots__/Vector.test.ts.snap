// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vector should add two Vectors and return a new Vector 1`] = `
Vector {
  "elements": Array [
    5,
    7,
    9,
  ],
  "length": 3,
}
`;

exports[`Vector should calculate the angle between two Vectors 1`] = `0.22572612855273394`;

exports[`Vector should calculate the angle between two Vectors in degrees 1`] = `12.93315449189912`;

exports[`Vector should calculate the bisect of two Vectors 1`] = `
Vector {
  "elements": Array [
    0.3638667954886596,
    0.5556981892812022,
    0.7475295830737448,
  ],
  "length": 3,
}
`;

exports[`Vector should calculate the cross product of two Vectors 1`] = `
Vector {
  "elements": Array [
    -3,
    6,
    -3,
  ],
  "length": 3,
}
`;

exports[`Vector should calculate the distance between two Vectors 1`] = `5.196152422706632`;

exports[`Vector should calculate the dot product of two Vectors 1`] = `32`;

exports[`Vector should calculate the outer product of two Vectors 1`] = `
Matrix {
  "columns": 3,
  "elements": Array [
    Array [
      4,
      5,
      6,
    ],
    Array [
      8,
      10,
      12,
    ],
    Array [
      12,
      15,
      18,
    ],
  ],
  "rows": 3,
}
`;

exports[`Vector should calculate the sight of a Vector 1`] = `true`;

exports[`Vector should calculate the skew symmetric matrix of a Vector 1`] = `
Matrix {
  "columns": 3,
  "elements": Array [
    Array [
      0,
      -3,
      2,
    ],
    Array [
      3,
      0,
      -1,
    ],
    Array [
      -2,
      1,
      0,
    ],
  ],
  "rows": 3,
}
`;

exports[`Vector should convert a Vector to a Vector3D 1`] = `
Vector3D {
  "x": 2,
  "y": 3,
  "z": undefined,
}
`;

exports[`Vector should create a Vector with a negative length 1`] = `[Function]`;

exports[`Vector should get a column from a Vector 1`] = `
Matrix {
  "columns": 1,
  "elements": Array [
    Array [
      1,
    ],
    Array [
      2,
    ],
    Array [
      3,
    ],
  ],
  "rows": 3,
}
`;

exports[`Vector should get a row from a Vector 1`] = `
Matrix {
  "columns": 3,
  "elements": Array [
    Array [
      1,
      2,
      3,
    ],
  ],
  "rows": 1,
}
`;

exports[`Vector should join two Vectors 1`] = `
Vector {
  "elements": Array [
    1,
    2,
    3,
    4,
    5,
    6,
  ],
  "length": 6,
}
`;

exports[`Vector should normalize a Vector and return a new Vector 1`] = `
Vector {
  "elements": Array [
    0.6000000000000001,
    0.8,
  ],
  "length": 2,
}
`;

exports[`Vector should return a list representation of a Vector 1`] = `
Array [
  1,
  2,
  3,
]
`;

exports[`Vector should return a string representation of a Vector 1`] = `"[1, 2, 3]"`;

exports[`Vector should return an array representation of a Vector 1`] = `
Float64Array [
  1,
  2,
  3,
]
`;

exports[`Vector should rotate a Vector around the x-axis 1`] = `
Vector {
  "elements": Array [
    1,
    3,
    -1.9999999999999998,
  ],
  "length": 3,
}
`;

exports[`Vector should rotate a Vector around the y-axis 1`] = `
Vector {
  "elements": Array [
    -3,
    2,
    1.0000000000000002,
  ],
  "length": 3,
}
`;

exports[`Vector should rotate a Vector around the z-axis 1`] = `
Vector {
  "elements": Array [
    2,
    -0.9999999999999999,
    3,
  ],
  "length": 3,
}
`;

exports[`Vector should slice a Vector 1`] = `
Vector {
  "elements": Array [
    2,
  ],
  "length": 1,
}
`;
